cpp/dolfinx/la/MatrixCSR.h

```diff
@@ -0,0 +1,491 @@
+      _unpack_pos.push_back(d);
+    }
+  }
+
+  /// Move constructor
+  /// @todo Check handling of MPI_Request
+  MatrixCSR(MatrixCSR&& A) = default;
+
+  /// Set all non-zero local entries to a value
+  /// @param[in] x The value to set non-zero matrix entries to
+  /// @todo This should probably also set ghost rows
+  void set(T x) { std::fill_n(_data.begin(), _index_maps[0]->size_local(), x); }
+
+  /// Set values in the matrix
+  /// @note Only entries included in the sparsity pattern used to
+  /// initialize the matrix can be set
+  /// @note All indices are local to the calling MPI rank
+  /// @param[in] x The `m` by `n` dense block of values (row-major) to
+  /// set in the matrix
+  /// @param[in] rows The row indices of `x`
+  /// @param[in] cols The column indices of `x`
+  void set(const xtl::span<const T>& x,
+           const xtl::span<const std::int32_t>& rows,
+           const xtl::span<const std::int32_t>& cols)
+  {
+    assert(x.size() == rows.size() * cols.size());
+    const std::int32_t local_size0 = _index_maps[0]->size_local();
+    const std::int32_t num_ghosts0 = _index_maps[0]->num_ghosts();
+    for (std::size_t r = 0; r < rows.size(); ++r)
+    {
+      // Columns indices for row
+      std::int32_t row = rows[r];
+
+      // Current data row
+      const T* xr = x.data() + r * cols.size();
+
+      if (row < local_size0)
+      {
+        auto cit0 = std::next(_cols.begin(), _row_ptr[row]);
+        auto cit1 = std::next(_cols.begin(), _row_ptr[row + 1]);
+        for (std::size_t c = 0; c < cols.size(); ++c)
+        {
+          // Find position of column index
+          auto it = std::lower_bound(cit0, cit1, cols[c]);
+          assert(it != cit1);
+          assert(*it == cols[c]);
+          std::size_t d = std::distance(_cols.begin(), it);
+          _data[d] = xr[c];
+        }
+      }
+      else
+        throw std::runtime_error("Local row out of range");
+    }
+  }
+
+  /// Accumulate values in the matrix
+  /// @note Only entries included in the sparsity pattern used to
+  /// initialize the matrix can be accumulated in to
+  /// @note All indices are local to the calling MPI rank
+  /// @param[in] x The `m` by `n` dense block of values (row-major) to
+  /// add to the matrix
+  /// @param[in] rows The row indices of `x`
+  /// @param[in] cols The column indices of `x`
+  void add(const xtl::span<const T>& x,
+           const xtl::span<const std::int32_t>& rows,
+           const xtl::span<const std::int32_t>& cols)
+  {
+    assert(x.size() == rows.size() * cols.size());
+    const std::int32_t local_size0 = _index_maps[0]->size_local();
+    const std::int32_t num_ghosts0 = _index_maps[0]->num_ghosts();
+    for (std::size_t r = 0; r < rows.size(); ++r)
+    {
+      // Columns indices for row
+      std::int32_t row = rows[r];
+
+      // Current data row
+      const T* xr = x.data() + r * cols.size();
+
+      if (row < local_size0 + num_ghosts0)
```

Comment:
This confusing. No need for a if/else - just put the check and exception at the top.